### ISSUE TITLE
Unsat Cores and Optimisation

### ISSUE THREAD

**Issue by iamrecursion** [Z3Prover/z3#1577]
Apparently the production of unsat cores is not supported in optimisation mode. I was wondering if this is a technical constraint or just an unimplemented feature as I've currently got a case where I have to run the solver a second time if the optimisation run produces unsat.

**Comment by iamrecursion** [comment]
That's good to know, thank you! I'm currently running the query again without optimisation if it fails so I can get the unsat core back out. 

I'd certainly be interested if optimisation mode could produce unsat cores, but it's not a critical need by any stretch of the imagination.

**Comment by LeventErkok** [comment]
Thanks @NikolajBjorner 

Would the following strategy be performant?

```
1.  Assert all constraints except optimization objectives
2.  Issue `check-sat`
        2.1  If `unsat`, get the unsat-core: done
        2.2 If `sat`:
              2.2.1 Assert the optimization objectives
              2.2.2 Issue `check-sat` again
              2.2.3 Get objective values
```

Would the second `check-sat` in step 2.2.2 benefit from all the learned-lemmas from the first in step 2; or would this be a "solve from scratch" since we now have optimization objectives?

**Comment by NikolajBjorner** [comment]
it wouldn't. It would solve from scratch. You can always run the to queries in parallel :-).

**Comment by LeventErkok** [comment]
Good to know. That's actually not a bad idea at all; the original context is created from within Haskell, which has excellent support for parallel execution.

I think it'd still be nice to have access to unsat-cores in optimization; but feel free to close this ticket if it doesn't rise up in the priority chain!

**Comment by LeventErkok** [comment]
@NikolajBjorner 

Thanks for fixing this Nikolaj! However, I'm not having much luck with it:

```
(set-option :produce-unsat-cores true)
(set-logic ALL)
(declare-fun i () Int)
(assert (! (> i 5) :named |c1|))
(assert (! (< i 1) :named |c2|))
(minimize i)
(check-sat)
(get-unsat-core)
```

produces:

```
unsat
()
```

If I remove the call to `minimize`, I get the expected:

```
unsat
(c1 c2)
```

Is `get-unsat-core` not the right call to make in the optimization context?

**Comment by LeventErkok** [comment]
Thanks @NikolajBjorner! Seems to be working now.

**Comment by iamrecursion** [comment]
Brilliant! Thanks so much @NikolajBjorner!
